pocs/observatory.py

```diff
@@ -537,6 +555,235 @@ def close_dome(self):
+                           az=None,
+                           min_counts=1000,
+                           max_counts=12000,
+                           bias=2048,
+                           max_exptime=60.,
+                           camera_list=None,
+                           target_adu_percentage=0.5,
+                           max_num_exposures=10,
+                           take_darks=False,
+                           *args, **kwargs
+                           ):
+        """Take flat fields.
+
+        This method will slew the mount to the given AltAz coordinates (which
+        should be roughly opposite of the setting sun) and then begin the flat-field
+        procedure. The first image starts with a simple 1 second exposure and
+        after each image is taken the average counts are anazlyzed and the exposure
+        time is adjusted to try to keep the counts close to `target_adu_percentage`
+        of the `(max_counts + min_counts) - bias`.
+
+        The next exposure time is calculated as:
+
+        ```
+            exp_time = int(previous_exp_time * (target_adu / counts) *
+                       (2.0 ** (elapsed_time / 180.0)) + 0.5)
+        ```
+
+        Under- and over-exposed images are rejected. If image is saturated with
+        a short exposure the method will wait 60 seconds before beginning next
+        exposure.
+
+        Optionally, the method can also take dark exposures of equal exposure
+        time to each flat-field image.
+
+
+        Args:
+            alt (float, optional): Altitude for flats
+            az (float, optional): Azimuth for flats
+            min_counts (int, optional): Minimum ADU count
+            max_counts (int, optional): Maximum ADU count
+            bias (int, optional): Default bias for the cameras
+            max_exptime (float, optional): Maximum exposure time before stopping
+            camera_list (list, optional): List of cameras to use for flat-fielding
+            target_adu_percentage (float, optional): Exposure time will be adjust so
+                that counts are close to: target * (`min_counts` + `max_counts`). Default
+                to 0.5
+            max_num_exposures (int, optional): Maximum number of flats to take
+            take_darks (bool, optional): If dark fields matching flat fields should be
+                taken, default False.
+            *args (TYPE): Description
+            **kwargs (TYPE): Description
+        """
+        target_adu = target_adu_percentage * (min_counts + max_counts)
+
+        image_dir = self.config['directories']['images']
+
+        flat_obs = self._create_flat_field_observation(alt=alt, az=az)
+
+        if camera_list is None:
+            camera_list = list(self.cameras.keys())
+
+        # Start with 1 second exposures for each camera
+        exp_times = {cam_name: [1. * u.second] for cam_name in camera_list}
+
+        # Loop until conditions are met for flat-fielding
+        while True:
```

Comment:
Nothing to do yet, but I'm worried about this being a potentially very long running function. Seems like something worthy of an async work flow system (ala a state machine).

Reply:
Correct, it currently blocks inside this loop for a while. There are a number of checks to get out of the loop although notably none for weather. Should be since at least at MLO we can have rapidly changing conditions right as sun goes down.